examples/cms/src/App.tsx

```diff
@@ -0,0 +1,86 @@
+import "./App.css"
+
+import { framer, ManagedCollection } from "framer-plugin"
+import { useEffect, useLayoutEffect, useState, useRef } from "react"
+import { DataSource, getDataSource, getDataSourcesIds } from "./data"
+import { FieldMapping } from "./FieldMapping"
+import { SelectDataSource } from "./SelectDataSource"
+import { Spinner } from "./components/Spinner"
+
+interface AppProps {
+    collection: ManagedCollection
+    previouslyConfiguredDataSourceId: string | null
+    previouslyConfiguredSlugFieldId: string | null
+}
+
+export function App({ collection, previouslyConfiguredDataSourceId, previouslyConfiguredSlugFieldId }: AppProps) {
+    const [dataSource, setDataSource] = useState<DataSource | null>(null)
+
+    const hasStartedLoadingDataSource = useRef(false)
+    const [isLoadingConfiguredDataSource, setIsLoadingConfiguredDataSource] = useState(
+        Boolean(previouslyConfiguredDataSourceId)
+    )
+    const hasDataSourceSelected = Boolean(isLoadingConfiguredDataSource || dataSource)
+
+    useLayoutEffect(() => {
+        framer.showUI({
+            width: hasDataSourceSelected ? 360 : 320,
+            height: hasDataSourceSelected ? 425 : 305,
+            minWidth: hasDataSourceSelected ? 360 : undefined,
+            minHeight: hasDataSourceSelected ? 425 : undefined,
+            resizable: dataSource !== null,
+        })
+    }, [hasDataSourceSelected, dataSource])
+
+    useEffect(() => {
+        if (!previouslyConfiguredDataSourceId || hasStartedLoadingDataSource.current) {
+            return
+        }
+
+        hasStartedLoadingDataSource.current = true
+        getDataSource(previouslyConfiguredDataSourceId)
+            .then(setDataSource)
+            .catch(error => {
+                console.error(error)
+                framer.notify(
+                    `Error loading previously configured data source "${previouslyConfiguredDataSourceId}". Check the logs for more details.`,
+                    {
+                        variant: "error",
+                    }
+                )
+            })
+            .finally(() => {
+                setIsLoadingConfiguredDataSource(false)
+            })
+    }, [previouslyConfiguredDataSourceId])
+
+    if (!isLoadingConfiguredDataSource && !dataSource) {
+        return <SelectDataSource dataSourceIds={getDataSourcesIds()} onSelectDataSource={setDataSource} />
+    }
+
+    if (isLoadingConfiguredDataSource) {
+        return <Spinner />
+    }
```

Comment:
By moving the rendering of `Spinner`  up before `SelectDataSource`, you don't need to check for `!isLoadingConfiguredDataSource`